File: 2020/11/week3/minhyungPark/algorithm/LeetCode1647.kt
/**
 * Leetcode 1647. Minimum Deletions to Make Character Frequencies Unique
 * https://leetcode.com/contest/weekly-contest-214/problems/minimum-deletions-to-make-character-frequencies-unique/
 */

class LeetCode1647 {
    fun minDeletions(s: String): Int {
        val frequency = IntArray(26)
        s.forEach { c -> frequency[c-'a']++ }
        println(frequency.contentToString())

        val nums = IntArray(frequency.max()!! + 1)
        frequency.filter { it!=0 }.forEach { nums[it]++ }

        var result = 0
        for (i in nums.size-1 downTo 1) {
            if (nums[i] >= 2) {
                result+= nums[i] - 1
                nums[i-1] += nums[i] -1
            }
        }
        return result
    }
}

File: 2020/11/week4/1ilsang/algorithm/leetcode1647.js
/**
 * https://leetcode.com/contest/weekly-contest-214/problems/minimum-deletions-to-make-character-frequencies-unique/
 * 1647. Minimum Deletions to Make Character Frequencies Unique
 * @param {string} s
 * @return {number}
 */
const minDeletions = (s) => {
  const arr = new Array(26).fill(0);
  let answer = 0;

  Array.from(s).forEach((e) => arr[e.charCodeAt() - 97]++);
  arr.sort((a, b) => b - a);

  for (let i = 1; i < 26; i++) {
    while (arr[i] && arr[i - 1] <= arr[i]) {
      arr[i]--;
      answer++;
    }
  }
  return answer;
};
